extension/src/documents.ts

```diff
@@ -50,46 +125,66 @@ export class DocumentManager{
 		}
 	}
 
-	// Gets an array of PRJ ReadonlyDocument references, but prefers vscode.TextDocument's as their source when available
+	// Gets an array of PRJ ReadonlyDocument references, but verifies the content is based on a vscode.TextDocument's when available
 	private getProjectDocuments(): ReadonlyDocument[] {
+		const result: ReadonlyDocument[] = [];
 		if (ProjectTreeProvider.hasProjectOpened()) {
-			return ProjectTreeProvider.instance().projectNode.sourceFiles.map(x => 
-				this._opened.has(x.document.fileName) ? ReadonlyDocument.getMemoryDocument(this._opened.get(x.document.fileName)) : x.document
-				);
-		} else {
-			return [];
+			this.projectKeys.forEach(key => {
+				if (this._cached.has(key)){
+					const sources = this._cached.get(key);
+					sources.flags.add(Origins.PROJECT);
+					if (sources.native && !sources.internal.equal(sources.native)){
```

Comment:
I think this 3 lines codes are duplicated several times, maybe extract to a function

Reply:
I know there is a lot of stuff that looks the same but sometimes has subtle differences. You are right, I should try a lot harder to clean this up with some meaningful function names even if they aren't recycled very much.

Reply:
I made 2.5 functions that consume the vast majority of the complexity. Should be a lot clearer now. These primary handlers do some unnecessary things/checks for certain circumstance, but I don't believe this hurts anything.